File: folkmq-lang-python/folkmq/folkmq/common/MqMetasResolverV1.py

from socketd import SocketD
from socketd.transport.core import Entity
from socketd.transport.core.Flags import Flags
from socketd.transport.core.Message import Message
from socketd.transport.core.entity.EntityDefault import EntityDefault
from socketd.transport.core.entity.MessageBuilder import MessageBuilder
from socketd.utils.StrUtils import StrUtils

from folkmq.client.MqMessage import MqMessage
from folkmq.common.MqConstants import MqConstants
from folkmq.common.MqMetasResolver import MqMetasResolver
from folkmq.common.MqMetasV1 import MqMetasV1
from folkmq.common.MqMetasV2 import MqMetasV2


class MqMetasResolverV1(MqMetasResolver):
    def version(self) -> int:
        return 1

    def get_sender(self, m: Entity) -> str:
        return m.meta_or_default(MqMetasV2.MQ_META_SENDER, "");

    def get_key(self, m: Entity) -> str:
        return m.meta_or_default(MqMetasV1.MQ_META_KEY, "");

    def get_tag(self, m: Entity) -> str:
        return m.meta_or_default(MqMetasV2.MQ_META_TAG, "");

    def get_topic(self, m: Entity) -> str:
        return m.meta_or_default(MqMetasV1.MQ_META_TOPIC, "");

    def get_consumer_group(self, m: Entity) -> str:
        return m.meta_or_default(MqMetasV1.MQ_META_CONSUMER_GROUP, "");

    def set_consumer_group(self, m: Entity, consumerGroup: str):
        m.put_meta(MqMetasV1.MQ_META_CONSUMER_GROUP, consumerGroup);

    def get_qos(self, m: Entity) -> int:
        return 0 if "0" == m.meta(MqMetasV1.MQ_META_QOS) else 1;

    def get_times(self, m: Entity) -> int:
        return int(m.meta_or_default(MqMetasV1.MQ_META_TIMES, "0"));

    def set_times(self, m: Entity, times: int):
        m.put_meta(MqMetasV1.MQ_META_TIMES, times.toString());

    def get_expiration(self, m: Entity) -> float:
        return float(m.meta_or_default(MqMetasV1.MQ_META_EXPIRATION, "0"));

    def set_expiration(self, m: Entity, expiration: int | None):
        if expiration is None:
            m.delMeta(MqMetasV1.MQ_META_EXPIRATION)
        else:
            m.put_meta(MqMetasV1.MQ_META_EXPIRATION, expiration.toString())

    def get_scheduled(self, m: Entity) -> int:
        return int(m.meta_or_default(MqMetasV1.MQ_META_SCHEDULED, "0"))

    def set_scheduled(self, m: Entity, scheduled: int):
        m.put_meta(MqMetasV1.MQ_META_SCHEDULED, scheduled.toString())

    def is_sequence(self, m: Entity) -> bool:
        return "1" == (m.meta_or_default(MqMetasV1.MQ_META_SEQUENCE, "0"))

    def is_broadcast(self, m: Entity) ->bool:
        return "1" == (m.meta(MqMetasV2.MQ_META_BROADCAST))

    def is_transaction(self, m: Entity) -> bool:
        return "1" == (m.meta(MqMetasV2.MQ_META_TRANSACTION))

    def set_transaction(self, m: Entity, isTransaction: bool):
        m.put_meta(MqMetasV2.MQ_META_TRANSACTION, ( "1" if isTransaction else "0"))

    def publish_entity_build(self, topic: str, message: MqMessage) -> EntityDefault:
        # 构建消息实体
        entity = EntityDefault().data_set(message.get_body())

        entity.meta_put(MqMetasV1.MQ_META_KEY, message.get_key())
        entity.meta_put(MqMetasV1.MQ_META_TOPIC, topic)
        entity.meta_put(MqMetasV1.MQ_META_QOS, ("0" if message.get_qos() == 0 else "1"))

        # 标签
        if message.get_tag():
            entity.meta_put(MqMetasV2.MQ_META_TAG, message.get_tag())


        # 定时派发
        if message.getScheduled():
            entity.meta_put(MqMetasV1.MQ_META_SCHEDULED, str(message.getScheduled().getTime()))
        else:
            entity.meta_put(MqMetasV1.MQ_META_SCHEDULED, "0")

        # 过期时间
        if message.get_expiration() is not None:
            entity.meta_put(MqMetasV1.MQ_META_EXPIRATION, str(message.get_expiration().getTime()))


        if  message.is_transaction():
            entity.meta_put(MqMetasV2.MQ_META_TRANSACTION, "1")


        if message.get_sender():
            entity.meta_put(MqMetasV2.MQ_META_SENDER, message.get_sender())

        if message.is_broadcast():
            entity.meta_put(MqMetasV2.MQ_META_BROADCAST, "1")

        # 是否有序
        if  message.is_sequence() or message.is_transaction():
            entity.at(MqConstants.PROXY_AT_BROKER_HASH)

        if message.is_sequence():
            entity.meta_put(MqMetasV1.MQ_META_SEQUENCE, "1")
        else:
            entity.at(MqConstants.PROXY_AT_BROKER)

        # 用户属性
        #message.getAttrMap().forEach((value, key, map) = > {
        #    entity.put_meta(MqConstants.MQ_ATTR_PREFIX + key, value)
        #})

        return entity

    def routing_message_build(self, topic: str, message: MqMessage) -> Message:
        entity = self.publish_entity_build(topic, message).at(MqConstants.PROXY_AT_BROKER)
        messageDefault =  MessageBuilder().flag(Flags.Message) \
                                .sid(StrUtils.guid()) \
                                .event(MqConstants.MQ_EVENT_PUBLISH) \
                                .entity(entity) \
                                .build()
        return messageDefault

File: folkmq-lang-java/folkmq/src/main/java/org/noear/folkmq/common/MqMetasResolverV1.java
package org.noear.folkmq.common;

import org.noear.folkmq.client.MqMessage;
import org.noear.socketd.transport.core.Entity;
import org.noear.socketd.transport.core.Flags;
import org.noear.socketd.transport.core.Message;
import org.noear.socketd.transport.core.MessageInternal;
import org.noear.socketd.transport.core.entity.EntityDefault;
import org.noear.socketd.transport.core.entity.MessageBuilder;
import org.noear.socketd.utils.StrUtils;

import java.util.Map;

/**
 * 消息元信息分析器 v1
 *
 * @author noear
 * @since 1.2
 */
public class MqMetasResolverV1 implements MqMetasResolver {
    @Override
    public int version() {
        return 1;
    }

    @Override
    public String getSender(Entity m) {
        return m.metaOrDefault(MqMetasV2.MQ_META_SENDER, "");
    }

    public String getKey(Entity m) {
        return m.metaOrDefault(MqMetasV1.MQ_META_KEY, "");
    }

    @Override
    public String getTag(Entity m) {
        return m.metaOrDefault(MqMetasV2.MQ_META_TAG, "");
    }

    public String getTopic(Entity m) {
        return m.metaOrDefault(MqMetasV1.MQ_META_TOPIC, "");
    }

    public String getConsumerGroup(Entity m) {
        return m.metaOrDefault(MqMetasV1.MQ_META_CONSUMER_GROUP, "");
    }

    public void setConsumerGroup(Entity m, String consumerGroup){
        m.putMeta(MqMetasV1.MQ_META_CONSUMER_GROUP, consumerGroup);
    }

    public int getQos(Entity m) {
        return "0".equals(m.meta(MqMetasV1.MQ_META_QOS)) ? 0 : 1;
    }

    public int getTimes(Entity m) {
        return Integer.parseInt(m.metaOrDefault(MqMetasV1.MQ_META_TIMES, "0"));
    }

    public void setTimes(Entity m, int times){
        m.putMeta(MqMetasV1.MQ_META_TIMES, String.valueOf(times));
    }

    @Override
    public long getExpiration(Entity m) {
        return Long.parseLong(m.metaOrDefault(MqMetasV1.MQ_META_EXPIRATION, "0"));
    }

    @Override
    public void setExpiration(Entity m, Long expiration) {
        if (expiration == null) {
            m.delMeta(MqMetasV1.MQ_META_EXPIRATION);
        } else {
            m.putMeta(MqMetasV1.MQ_META_EXPIRATION, expiration.toString());
        }
    }

    @Override
    public long getScheduled(Entity m) {
        return Long.parseLong(m.metaOrDefault(MqMetasV1.MQ_META_SCHEDULED, "0"));
    }

    @Override
    public void setScheduled(Entity m, long scheduled){
        m.putMeta(MqMetasV1.MQ_META_SCHEDULED, String.valueOf(scheduled));
    }

    @Override
    public boolean isSequence(Entity m) {
        return "1".equals(m.metaOrDefault(MqMetasV1.MQ_META_SEQUENCE, "0"));
    }

    @Override
    public boolean isBroadcast(Entity m) {
        return "1".equals(m.meta(MqMetasV2.MQ_META_BROADCAST));
    }

    @Override
    public boolean isTransaction(Entity m) {
        return "1".equals(m.meta(MqMetasV2.MQ_META_TRANSACTION));
    }

    @Override
    public void setTransaction(Entity m, boolean isTransaction) {
        m.putMeta(MqMetasV2.MQ_META_TRANSACTION, (isTransaction ? "1" : "0"));
    }


    /**
     * 发布实体构建
     *
     * @param topic   主题
     * @param message 消息
     */
    public EntityDefault publishEntityBuild(String topic, MqMessage message) {
        //构建消息实体
        EntityDefault entity = new EntityDefault().dataSet(message.getBody());

        entity.metaPut(MqMetasV1.MQ_META_KEY, message.getKey());
        entity.metaPut(MqMetasV1.MQ_META_TOPIC, topic);
        entity.metaPut(MqMetasV1.MQ_META_QOS, (message.getQos() == 0 ? "0" : "1"));

        //标签
        if (StrUtils.isNotEmpty(message.getTag())) {
            entity.metaPut(MqMetasV2.MQ_META_TAG, message.getTag());
        }

        //定时派发
        if (message.getScheduled() == null) {
            entity.metaPut(MqMetasV1.MQ_META_SCHEDULED, "0");
        } else {
            entity.metaPut(MqMetasV1.MQ_META_SCHEDULED, String.valueOf(message.getScheduled().getTime()));
        }

        //过期时间
        if (message.getExpiration() != null) {
            entity.metaPut(MqMetasV1.MQ_META_EXPIRATION, String.valueOf(message.getExpiration().getTime()));
        }

        if (message.isTransaction()) {
            entity.metaPut(MqMetasV2.MQ_META_TRANSACTION, "1");
        }

        if(StrUtils.isNotEmpty(message.getSender())){
            entity.metaPut(MqMetasV2.MQ_META_SENDER, message.getSender());
        }

        if(message.isBroadcast()){
            entity.metaPut(MqMetasV2.MQ_META_BROADCAST, "1");
        }

        //是否有序
        if (message.isSequence() || message.isTransaction()) {
            entity.at(MqConstants.BROKER_AT_SERVER_HASH);

            if (message.isSequence()) {
                entity.metaPut(MqMetasV1.MQ_META_SEQUENCE, "1");
            }
        } else {
            entity.at(MqConstants.BROKER_AT_SERVER);
        }

        //用户属性
        for (Map.Entry<String, String> kv : message.getAttrMap().entrySet()) {
            entity.putMeta(MqConstants.MQ_ATTR_PREFIX + kv.getKey(), kv.getValue());
        }

        return entity;
    }

    /**
     * 路由消息构建
     *
     * @param topic   主题
     * @param message 消息
     */
    public Message routingMessageBuild(String topic, MqMessage message) {
        Entity entity = publishEntityBuild(topic, message)
                .at(MqConstants.BROKER_AT_SERVER);

        MessageInternal messageDefault = new MessageBuilder()
                .flag(Flags.Message)
                .sid(StrUtils.guid())
                .event(MqConstants.MQ_EVENT_PUBLISH)
                .entity(entity)
                .build();

        return messageDefault;
    }
}